### PR TITLE
Adapt adapter pattern for product load

### PR DESCRIPTION
**Pull request by dunqan**
Use Connectors, Adapters and Normalizers for loading product data

Closes GH-1980

### CODE REVIEW

File: projects/core/src/product/occ/converters/product-image-normalizer.ts [hunk 1]
```diff
@@ -1,12 +1,25 @@
...
+import { OccConfig } from '../../../occ';
 import { Product } from '../../../occ/occ-models/occ.models';
+import { Converter } from '../../../util/converter.service';
 
-@Injectable()
-export class ProductImageConverterService {
+@Injectable({
+  providedIn: 'root',
+})
+export class ProductImageNormalizer implements Converter<Product, Product> {
```
Comment (tobi-or-not-tobi) [inline projects/core/src/product/occ/converters/product-image-normalizer.ts]: Can we add the UI model here? The backend model has a single list of images, while we like to have a more semantic structure in the UI model. We do not have this reflected in the UI model though.

File: projects/core/src/product/connectors/product/product.adapter.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { Observable } from 'rxjs';
+import { Product } from '../../../occ/occ-models/occ.models';
+
+export abstract class ProductAdapter {
+  abstract load(productCode: string): Observable<Product>;
```
Comment (tobi-or-not-tobi) [inline projects/core/src/product/connectors/product/product.adapter.ts]: can we add a bit of documentation for this abstract method?

Comment (dunqan) [reply projects/core/src/product/connectors/product/product.adapter.ts]: 👍

File: projects/core/src/product/occ/converters/product-reference-normalizer.ts [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 
 import { Product } from '../../../occ/occ-models/occ.models';
+import { Converter } from '../../../util/converter.service';
 
 @Injectable()
-export class ProductReferenceConverterService {
-  convertProduct(product: Product): void {
-    if (product.productReferences) {
-      product.productReferences = this.populate(product.productReferences);
+export class ProductReferenceNormalizer implements Converter<Product, Product> {
```
[the remark this answers is not in the document]
Comment (dunqan) [reply projects/core/src/product/occ/converters/product-reference-normalizer.ts]: ok, done

File: projects/core/src/product/occ/converters/product-reference-normalizer.ts [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
+  convert(source: Product, target?: Product): Product {
+    if (target === undefined) {
+      target = { ...source };
     }
+    if (source.productReferences) {
+      target.productReferences = this.populate(source.productReferences);
```
[the remark this answers is not in the document]
Comment (dunqan) [reply projects/core/src/product/occ/converters/product-reference-normalizer.ts]: 👍

File: projects/core/src/product/occ/converters/product-image-normalizer.ts [hunk 5]
```diff
@@ -1,12 +1,25 @@
...
   constructor(protected config: OccConfig) {}
 
+  convert(source: Product, target?: Product): Product {
+    if (target === undefined) {
+      target = { ...source };
+    }
+    if (source.images) {
+      target.images = this.populate(source.images);
```
[the remark this answers is not in the document]
Comment (dunqan) [reply projects/core/src/product/occ/converters/product-image-normalizer.ts]: 👍

**Pull request merged**